aws/resource_aws_cloudhsm2_cluster.go

```diff
@@ -0,0 +1,384 @@
+	return &schema.Resource{
+		Create: resourceAwsCloudHsm2ClusterCreate,
+		Read:   resourceAwsCloudHsm2ClusterRead,
+		Update: resourceAwsCloudHsm2ClusterUpdate,
+		Delete: resourceAwsCloudHsm2ClusterDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceAwsCloudHsm2ClusterImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(120 * time.Minute),
+			Update: schema.DefaultTimeout(120 * time.Minute),
+			Delete: schema.DefaultTimeout(120 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"backup_identifier": {
+				Type:     schema.TypeString,
+				Computed: false,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"hsm_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateCloudHsm2HsmType,
+			},
+
+			"subnet_ids": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+				ForceNew: true,
+			},
+
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_certificates": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"cluster_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"cluster_csr": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"aws_hardware_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"hsm_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"manufacturer_hardware_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCloudHsm2ClusterImport(
+	d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	d.Set("cluster_id", d.Id())
+	return []*schema.ResourceData{d}, nil
+}
+
+func describeCloudHsm2Cluster(clusterId string, meta interface{}) (*cloudhsmv2.Cluster, error) {
+	conn := meta.(*AWSClient).cloudhsmv2conn
+
+	filters := []*string{&clusterId}
+	result := int64(1)
+	out, err := conn.DescribeClusters(&cloudhsmv2.DescribeClustersInput{
+		Filters: map[string][]*string{
+			"clusterIds": filters,
+		},
+		MaxResults: &result,
+	})
+	if err != nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster (%s) when waiting: %s", clusterId, err)
+		return nil, err
+	}
+
+	var cluster *cloudhsmv2.Cluster
+
+	for _, c := range out.Clusters {
+		if *c.ClusterId == clusterId {
```

Comment:
To prevent potential `nil` panics, we should use the SDK provided helper function here: `if aws.StringValue(c.ClusterId) == clusterId {`